Public/Src/Tools/Execution.Analyzer/Analyzers.Partner/ContentPlacementAnalyzer.cs

```diff
@@ -0,0 +1,532 @@
+
+            }
+            return new ContentPlacementAnalyzer(GetAnalysisInput(), sampleSize, sampleSizeHardLimit, buildQueue, buildId, buildStartTicks, buildDurationMs) { };
+        }
+
+        private static void WriteContentPlacementAnalyzerHelp(HelpWriter writer)
+        {
+            writer.WriteBanner("Content Placement Analyzer");
+            writer.WriteModeOption(nameof(AnalysisMode.ContentPlacement), "This analyzer parses a whole build (master perspective) and outputs content/workload-related data for machine learning purposes. It should only be run at the master node.");
+            writer.WriteOption("sampleProportion", "Required. The number of artifacts that will be analyzed/saved/written. This number is a minimun, since we try to make sure to take at least one file of each extension.", shortName: "sp");
+            writer.WriteOption("buildQueue", "Optional. The build queue in which this build ran. If not set, its necessary to get it when processing the output.", shortName: "bq");
+            writer.WriteOption("buildId", "Optional. The BuildId for this build. If not set, its necessary to get it when processing the output.", shortName: "bid");
+            writer.WriteOption("buildStartTicks", "Optional. The time (ticks) when the build started. If not set, its necessary to get it when processing the output.", shortName: "bst");
+            writer.WriteOption("buildDurationMs", "Optional. The duration of this build (milliseconds). If not set, its necessary to get it when processing the output.", shortName: "bd");
+            writer.WriteOption("sampleCountHardLimit", "Optional. This is a hard limit on the number of samples. We prioritize getting at least a file from each extension, but with this cap we cannot ensure that. It defaults to universe_size/2", shortName: "schl");
+
+        }
+    }
+
+    /// <summary>
+    /// Analyzer used to get stats on events (count and total size)
+    /// </summary>
+    internal sealed class ContentPlacementAnalyzer : Analyzer
+    {
+
+        private readonly string m_relativeOutputFile = "CpResults.json";
+        private readonly Dictionary<uint, IList<PipExecutionPerformanceEventData>> m_scheduledPipExecutionData = new Dictionary<uint, IList<PipExecutionPerformanceEventData>>(capacity: 10 * 1000);
+        private readonly Dictionary<uint, int> m_scheduledPipDependencyCount = new Dictionary<uint, int>(capacity: 10 * 1000);
+        private readonly Dictionary<uint, int> m_scheduledPipInputCount = new Dictionary<uint, int>(capacity: 10 * 1000);
+        private readonly Dictionary<uint, int> m_scheduledPipOutputCount = new Dictionary<uint, int>(capacity: 10 * 1000);
+        private readonly Dictionary<FileArtifact, FileContentInfo> m_fileContentMap = new Dictionary<FileArtifact, FileContentInfo>(capacity: 10 * 1000);
+        private readonly Dictionary<FileArtifact, IList<Pip>> m_artifactInputPips = new Dictionary<FileArtifact, IList<Pip>>(capacity: 10 * 1000);

```

Comment:
>ew Dictionary<FileArtifact, IList<Pi [](start = 85, length = 36)

We do have a helper class in BuildXL.Collections called: `MultiValueDictionary` that should simplify adding items quite a bit.